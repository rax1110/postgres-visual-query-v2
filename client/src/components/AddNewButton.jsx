import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import React from 'react';
import * as PropTypes from 'prop-types';
import { Button } from 'reactstrap';

export const AddNewButton = ({ id, size, onClick }) => (
  <Button
    className=""
    outline
    color="info"
    size={size}
    id={id}
    onClick={onClick}
  >
    <FontAwesomeIcon icon="plus" />
  </Button>
);

AddNewButton.propTypes = {
  id: PropTypes.string,
  size: PropTypes.string,
  onClick: PropTypes.func,
};
