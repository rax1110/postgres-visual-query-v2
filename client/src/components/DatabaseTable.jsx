import React from 'react';
import { Button, Tooltip } from 'reactstrap';
import { connect } from 'react-redux';
import _ from 'lodash';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import * as PropTypes from 'prop-types';
import { addColumn, addTable, removeTable } from '../actions/queryActions';
import { withToggle } from '../hocs/withToggle';

export const DatabaseTable = ({ data, checked, id, addTableProp, toggle, toggleStatus }) => {
  const handleOnClick = () => {
    addTableProp(data);
  };

  let tableTypeColor = 'primary';

  if (_.isEqual(data.table_type, 'VIEW')) {
    tableTypeColor = 'info';
  } else if (_.isEqual(data.table_type, 'FOREIGN')) {
    tableTypeColor = 'secondary';
  } else if (_.isEqual(data.table_type, 'MATERIALIZED VIEW')) {
    tableTypeColor = 'light';
  }

  const btnSelected = checked ? 'success' : tableTypeColor;

  const modifiers = {
    preventOverflow: {
      enabled: false,
    },
    hide: {
      enabled: false,
    },
  };

  return (
    <div className="w-100 pr-1">
      <Button
        size="sm"
        color={btnSelected}
        id={id}
        className="btn-block my-1 pt-0 text-left"
        onClick={handleOnClick}
      >
        <small
          color={tableTypeColor}
          className="text-truncate align-self-start"
        >
          <span className="mr-1 px-0">
            <FontAwesomeIcon icon="th-large" />
          </span>
          {data.table_type}
        </small>
        <div className="text-truncate">
          {` ${data.table_name}`}
        </div>
      </Button>
      <Tooltip
        placement="right"
        isOpen={toggleStatus}
        target={id}
        toggle={toggle}
        modifiers={modifiers}
        delay={{ show: 200, hide: 0 }}
        className=""
      >
        {data.table_name}
      </Tooltip>
    </div>
  );
};

DatabaseTable.propTypes = {
  data: PropTypes.shape({ table_type: PropTypes.string, table_name: PropTypes.string }),
  checked: PropTypes.bool,
  id: PropTypes.string,
  addTableProp: PropTypes.func,
  toggle: PropTypes.func,
  toggleStatus: PropTypes.bool,
};

const mapDispatchToProps = {
  addColumn,
  addTableProp: data => addTable(data),
  removeTable,
};

export default withToggle(connect(null, mapDispatchToProps)(DatabaseTable));
