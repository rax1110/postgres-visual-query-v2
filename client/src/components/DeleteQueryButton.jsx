import React from 'react';
import { Button } from 'reactstrap';
import { connect } from 'react-redux';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import * as PropTypes from 'prop-types';
import { deleteQuery, setActiveQuery } from '../actions/queryActions';
import { deleteQueries, removeMainFromQueries } from '../actions/queriesActions';

export const DeleteQueryButton = (props) => {
  const handleOnClick = () => {
    props.deleteQuery();

    const mainQuery = props.queries.find(query => query.id === 0);

    if (mainQuery) {
      props.removeMainFromQueries();
      props.setActiveQuery(mainQuery);
    } else {
      props.deleteQueries();
    }
  };

  return (
    <Button color="danger" className="mr-2" onClick={handleOnClick}>
      <FontAwesomeIcon icon="trash" />
    </Button>
  );
};

DeleteQueryButton.propTypes = {
  deleteQuery: PropTypes.func,
  deleteQueries: PropTypes.func,
  queries: PropTypes.arrayOf(PropTypes.shape({})),
  removeMainFromQueries: PropTypes.func,
  setActiveQuery: PropTypes.func,
};

const mapStateToProps = store => ({
  queries: store.queries,
});

const mapDispatchToProps = {
  deleteQuery,
  deleteQueries,
  removeMainFromQueries,
  setActiveQuery,
};

export default connect(mapStateToProps, mapDispatchToProps)(DeleteQueryButton);
