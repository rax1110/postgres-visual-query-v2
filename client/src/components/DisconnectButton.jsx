import React from 'react';
import { connect } from 'react-redux';
import { Button } from 'reactstrap';
import { Redirect } from 'react-router-dom';
import * as PropTypes from 'prop-types';
import { disconnect } from '../actions/hostActions';
import { translations } from '../utils/translations';

export const DisconnectButton = (props) => {
  const handleOnClick = () => {
    props.disconnect();
  };

  if (!props.connected) {
    return <Redirect to="/" />;
  }

  return (
    <Button size="lg" className="btn-block my-2 px-2" color="danger" onClick={handleOnClick}>
      {translations[props.language.code].sideBar.disconnectB}
    </Button>
  );
};

DisconnectButton.propTypes = {
  disconnect: PropTypes.func,
  connected: PropTypes.bool,
  language: PropTypes.shape({ code: PropTypes.string }),
};

const mapStateToProps = store => ({
  connected: store.host.connected,
  language: store.settings.language,
});

const mapDispatchToProps = {
  disconnect,
};

export default connect(mapStateToProps, mapDispatchToProps)(DisconnectButton);
