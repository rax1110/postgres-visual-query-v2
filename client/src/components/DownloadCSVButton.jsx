import React from 'react';
import { connect } from 'react-redux';
import { CSVLink } from 'react-csv';
import _ from 'lodash';
import { Button } from 'reactstrap';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import * as PropTypes from 'prop-types';

export const DownloadCSVButton = (props) => {
  const getHeaders = (result) => {
    if (_.isNull(result)) {
      return [];
    }

    const headers = [];

    result.fields.forEach((field) => {
      const header = {
        label: field.name,
        key: field.name,
      };

      headers.push(header);
    });

    return headers;
  };

  const getData = (result) => {
    if (_.isNull(result)) {
      return [];
    }
    return result.rows;
  };

  const disabled = _.isNull(props.result);

  return (
    <CSVLink
      className="mr-2"
      data={getData(props.result)}
      headers={getHeaders(props.result)}
      filename="result.csv"
    >
      <Button disabled={disabled}>
        <FontAwesomeIcon icon="download" />
        <div className="d-inline"> CSV</div>
      </Button>
    </CSVLink>
  );
};

DownloadCSVButton.propTypes = {
  result: PropTypes.shape({}),
};

const mapStateToProps = store => ({
  result: store.query.result,
});

export default connect(mapStateToProps)(DownloadCSVButton);
