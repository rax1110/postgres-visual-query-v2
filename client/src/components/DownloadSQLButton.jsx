import React from 'react';
import { connect } from 'react-redux';
import { Button } from 'reactstrap';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import * as PropTypes from 'prop-types';

export const DownloadSQLButton = (props) => {
  const downloadContent = (name, content) => {
    if (navigator.msSaveBlob) {
      const blobObject = new Blob([content], { type: 'text/plain' });

      window.navigator.msSaveOrOpenBlob(blobObject, name);
    } else {
      const aTag = document.createElement('a');
      const file = new Blob([content], { type: 'text/plain' });

      aTag.href = URL.createObjectURL(file);
      aTag.download = name;

      document.body.appendChild(aTag);
      aTag.click();
    }
  };

  return (
    <Button
      className="mr-2"
      onClick={() => downloadContent('select_query.sql', props.sql)}
    >
      <FontAwesomeIcon icon="download" />
      <div className="d-inline"> SQL</div>
    </Button>
  );
};

DownloadSQLButton.propTypes = {
  sql: PropTypes.string,
};

const mapStateToProps = store => ({
  sql: store.query.sql,
});

export default connect(mapStateToProps)(DownloadSQLButton);
