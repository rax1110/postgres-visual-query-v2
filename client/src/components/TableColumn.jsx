import React, { Component } from 'react';
import { connect } from 'react-redux';
import { Button, ButtonGroup, Tooltip } from 'reactstrap';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import _ from 'lodash';
import * as PropTypes from 'prop-types';
import TableColumnPopover from './TableColumnPopover';
import { addColumn, removeColumn } from '../actions/queryActions';
import { withToggle } from '../hocs/withToggle';
import { iconPicker } from '../utils/iconPicker';

export class TableColumn extends Component {
  constructor(props) {
    super(props);

    this.state = {
      target: props.id,
      foreignKeys: this.findForeignKeys(),
    };

    this.handleRemove = this.handleRemove.bind(this);
    this.handleOnChange = this.handleOnChange.bind(this);
    this.findForeignKeys = this.findForeignKeys.bind(this);
  }

  handleRemove(data) {
    this.props.removeColumn(data);
  }

  handleOnChange() {
    this.props.addColumn(this.props.data);
  }

  findForeignKeys() {
    return this.props.data.constraints
      .filter(constraint => constraint.constraint_type.localeCompare('FOREIGN KEY') === 0);
  }

  render() {
    const btnSelected = this.props.columns.some(column => _.isEqual(column.table_id,
      this.props.data.table_id) && _.isEqual(column.column_name,
      this.props.data.column_name)) ? 'success' : 'light';

    const modifiers = {
      preventOverflow: {
        enabled: false,
      },
      hide: {
        enabled: false,
      },
    };

    return (
      <div className="d-flex align-items-center">
        <ButtonGroup size="sm" className="btn-block my-1 p-0 px-1">
          <Button
            color={btnSelected}
            className="text-left d-flex flex-row align-items-center w-100 px-1 border"
            id={`${this.state.target}-type`}
            onClick={this.handleOnChange}
          >
            {this.props.data.constraints.some(c => c.constraint_type.localeCompare('PRIMARY KEY') === 0)
            && <div className="mr-2 px-2 bg-info text-light rounded-pill">PK</div>}
            <div className="text-truncate d-flex">
              {!this.props.data.constraints.some(c => c.constraint_type.localeCompare('PRIMARY KEY') === 0)
              && (
                <span className={`mr-1 px-2 ${btnSelected === 'light' ? 'text-info' : 'text-light'}`}>
                  <FontAwesomeIcon icon={iconPicker(this.props.data.data_type)} />
                </span>
              )}
              <div className="text-truncate">{this.props.data.column_name}</div>
            </div>
            <div className="ml-auto pl-3">
              <div className="bg-light rounded-pill">
                {this.props.joins.map(join => join.conditions.map(condition => ((_.isEqual(
                  join.main_table.id,
                  this.props.data.table_id,
                ) && _.isEqual(condition.main_column, this.props.data.column_name))
                  || (_.isEqual(condition.secondary_table.id, this.props.data.table_id)
                    && _.isEqual(condition.secondary_column, this.props.data.column_name)))
                  && (
                  <FontAwesomeIcon
                    key={`condition-icon-${this.props.data.table_schema}-${this.props.data.table_name}-${this.props.data.column_name}-${_.uniqueId()}`}
                    className="mx-1"
                    icon="link"
                    style={{ color: join.color }}
                  />
                  )))}
              </div>
            </div>
          </Button>
          <Tooltip
            placement="left"
            isOpen={this.props.toggleStatus}
            target={`${this.state.target}-type`}
            toggle={this.props.toggle}
            modifiers={modifiers}
            delay={{ hide: 0 }}
          >
            {this.props.data.data_type}
          </Tooltip>
          {this.state.foreignKeys.length > 0
          && (
            <Button outline color="info" id={this.state.target} type="button">
              <FontAwesomeIcon
                icon="external-link-square-alt"
              />
            </Button>
          )}
        </ButtonGroup>
        {this.state.foreignKeys.length > 0
        && <TableColumnPopover target={this.state.target} foreignKeys={this.state.foreignKeys} />
        }
      </div>
    );
  }
}

TableColumn.propTypes = {
  id: PropTypes.string,
  removeColumn: PropTypes.func,
  addColumn: PropTypes.func,
  data: PropTypes.shape({
    constraints: PropTypes.array,
    table_id: PropTypes.number,
    column_name: PropTypes.string,
    table_schema: PropTypes.string,
    table_name: PropTypes.string,
    data_type: PropTypes.string,
  }),
  columns: PropTypes.arrayOf(PropTypes.shape({})),
  joins: PropTypes.arrayOf(PropTypes.shape({})),
  toggle: PropTypes.func,
  toggleStatus: PropTypes.bool,
};

const mapStateToProps = store => ({
  joins: store.query.joins,
  columns: store.query.columns,
});

const mapDispatchToProps = {
  removeColumn,
  addColumn,
};

export default withToggle(connect(mapStateToProps, mapDispatchToProps)(TableColumn));
