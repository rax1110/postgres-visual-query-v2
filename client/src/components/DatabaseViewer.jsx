import React from 'react';
import { connect } from 'react-redux';
import { Scrollbars } from 'react-custom-scrollbars';
import _ from 'lodash';
import * as PropTypes from 'prop-types';
import DatabaseTable from './DatabaseTable';
import { filterTable } from '../utils/filterTable';

export const DatabaseViewer = (props) => {
  const constructData = (table) => {
    const data = {
      table_schema: table.table_schema,
      table_name: table.table_name,
      table_type: table.table_type,
      table_alias: '',
    };

    let constraints = JSON.parse(JSON.stringify(props.constraints));

    constraints = constraints.filter(constraint => constraint.table_schema === data.table_schema
      && constraint.table_name === data.table_name);

    let columns = JSON.parse(JSON.stringify(props.columns));

    columns = columns.filter(column => column.table_name === data.table_name
      && column.table_schema === data.table_schema).map((column) => {
      const col = column;

      col.constraints = constraints.filter(
        constraint => _.includes(constraint.column_name, column.column_name),
      );

      delete col.table_name;
      delete col.table_schema;
      return col;
    });

    data.columns = columns;

    return data;
  };

  return (
    <div className="flex-fill">
      <Scrollbars className="d-flex" autoHide>
        <div className="mt-1 pr-2">
          {props.tables.map((table, index) => {
            const checked = props.queryTable.some(
              queryTable => _.isEqual(table.table_name, queryTable.table_name)
                && _.isEqual(table.table_schema, queryTable.table_schema),
            );

            const id = `database-table-${index}`;

            return table.table_schema === props.selectedSchema
              && filterTable(table, props.searchExpr)
              && (
                <DatabaseTable
                  data={constructData(table)}
                  checked={checked}
                  key={id}
                  id={id}
                />
              );
          })}
        </div>
      </Scrollbars>
    </div>
  );
};

DatabaseViewer.propTypes = {
  data: PropTypes.shape({ table_type: PropTypes.string, table_name: PropTypes.string }),
  constraints: PropTypes.arrayOf(PropTypes.shape({})),
  columns: PropTypes.arrayOf(PropTypes.shape({})),
  tables: PropTypes.arrayOf(PropTypes.shape({})),
  queryTable: PropTypes.arrayOf(PropTypes.shape({})),
  selectedSchema: PropTypes.string,
  searchExpr: PropTypes.string,
};

const mapStateToProps = store => ({
  tables: store.database.tables,
  schemas: store.database.schemas,
  selectedSchema: store.database.selectedSchema,
  constraints: store.database.constraints,
  columns: store.database.columns,
  queryTable: store.query.tables,
  searchExpr: store.database.searchExpr,
});

export default connect(mapStateToProps)(DatabaseViewer);
