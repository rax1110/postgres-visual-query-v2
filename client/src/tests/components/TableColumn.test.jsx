import React from 'react';
import { shallow } from 'enzyme';
import { TableColumn } from '../../components/TableColumn';

describe('Component: TableColumn', () => {
  let component;
  let props;

  beforeEach(() => {
    props = {
      removeColumn: jest.fn(),
      addColumn: jest.fn(),
      key: 'table-column-1',
      id: '1-table-column-1',
      data: {
        column_aggregate: '',
        column_alias: '',
        column_distinct_on: false,
        column_filter: '',
        column_group_by: false,
        column_name: 'nimetus',
        column_order: false,
        column_order_dir: true,
        constraints: [],
        data_type: 'smallint',
        display_in_query: true,
        id: 2,
        ordinal_position: 2,
        table_alias: '',
        table_id: 1,
        table_name: 'amet',
        table_schema: 'public',
      },
      joins: [{
        color: '#ddbf4f',
        conditions: [],
        id: 0,
        main_table: {
          table_alias: '',
          table_name: '',
          table_schema: '',
          columns: [
            {
              column_name: 'asukoht_kood',
              constraints: [],
              data_type: 'smallint',
              ordinal_position: 1,
              table_alias: '',
              table_id: 2,
              table_name: 'asukoht',
              table_schema: 'public',
            },
            {
              column_name: 'nimetus',
              constraints: [],
              data_type: 'smallint',
              ordinal_position: 2,
              table_alias: '',
              table_id: 2,
              table_name: 'asukoht',
              table_schema: 'public',
            },
          ],
        },
        type: 'inner',
      }],
      columns: [
        {
          column_aggregate: '',
          column_alias: '',
          column_distinct_on: false,
          column_filter: '',
          column_group_by: false,
          column_name: 'amet_kood',
          column_order: false,
          column_order_dir: true,
          constraints: [{
            column_name: '{amet_kood}',
            constraint_name: 'pk_amet',
            constraint_type: 'PRIMARY KEY',
            foreign_column_name: null,
            foreign_table_name: null,
            foreign_table_schema: null,
            table_name: 'amet',
            table_schema: 'public',
          }],
          data_type: 'smallint',
          display_in_query: true,
          id: 1,
          ordinal_position: 1,
          table_alias: '',
          table_id: 1,
          table_name: 'amet',
          table_schema: 'public',
        },
        {
          column_aggregate: '',
          column_alias: '',
          column_distinct_on: false,
          column_filter: '',
          column_group_by: false,
          column_name: 'nimetus',
          column_order: false,
          column_order_dir: true,
          constraints: [],
          data_type: 'smallint',
          display_in_query: true,
          id: 2,
          ordinal_position: 2,
          table_alias: '',
          table_id: 1,
          table_name: 'amet',
          table_schema: 'public',
        },
      ],
    };
  });

  test('TableColumn renders with default props', () => {
    component = shallow(<TableColumn {...props} />);

    expect(component).toMatchSnapshot();
  });

  test('TableColumn handleRemove calls removeColumn once', () => {
    component = shallow(<TableColumn {...props} />);

    component.instance().handleRemove({});

    expect(props.removeColumn.mock.calls.length).toBe(1);
  });

  test('TableColumn handleOnChange calls addColumn once', () => {
    component = shallow(<TableColumn {...props} />);

    component.instance().handleOnChange({});

    expect(props.addColumn.mock.calls.length).toBe(1);
  });
});
