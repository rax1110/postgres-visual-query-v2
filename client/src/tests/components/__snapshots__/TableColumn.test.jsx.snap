// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component: TableColumn TableColumn renders with default props 1`] = `ShallowWrapper {}`;
